support backend point weight for maglev

lrouter opened the pull request:
@brb Cilium does not support backend point weight of a service for maglev hash. This patch adds support it. It rebased the following pull request from master:
https://github.com/cilium/cilium/pull/13189

Now it lacks weight in service spec. So we could not specify backend point weight when creating a service. It would be finished in the future.

Commit:
service: support weighted backend points for maglev hash

Backend point of a service could not have its weight before.
This patch supports backend weight for maglev hash algorithm.
When a service contains some backend points with different
weights, flows destinated to service would be forwarded a
backend by its weight ratio.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

christarazi commented on pkg/maglev/maglev.go:
```diff
@@ -17,8 +17,8 @@ package maglev
 	"fmt"
-
 	"github.com/cilium/cilium/pkg/murmur3"
+	"sort"
```
Please reorder the imports as such:

```
import (
  <go std library>

  <cilium packages>
  
  <external packages>
)
```

lrouter replied:
done!

christarazi commented on pkg/service/service.go:
```diff
@@ -16,6 +16,7 @@ package service
 
 import (
 	"fmt"
+	"github.com/cilium/cilium/pkg/maglev"
```
Here as well

lrouter replied:
done!

christarazi commented on pkg/maglev/maglev.go:
```diff
@@ -74,33 +74,58 @@ func getPermutation(backends []string, m uint64) []uint64 {
 	return perm
 }
 
+//BackendPoint is a backend point with weight
```
Could you add a space like so to stay consistent with the rest of the file? There are many occurrences of this. 

```suggestion
// BackendPoint is a backend point with weight
```

lrouter replied:
done !

christarazi commented on pkg/maglev/maglev.go:
```diff
@@ -74,33 +74,58 @@ func getPermutation(backends []string, m uint64) []uint64 {
-func GetLookupTable(backends []string, m uint64) []int {
-	if len(backends) == 0 {
-		return nil
+func GetLookupTable(backends map[string]*BackendPoint, m uint64, maglevBackendIDsBuffer []uint16) {
```
Could you add a description to the godoc on what `maglevBackendIDsBuffer` is?

christarazi commented on pkg/maglev/maglev.go:
```diff
@@ -74,33 +74,58 @@ func getPermutation(backends []string, m uint64) []uint64 {
+			//Here (at fb) what we do, is using logic in control plane of balancer which is working like this:
+			//	1.Calculate sum of all real. (Eg 1+10)
+			//	2.Devide hash ring size by this sum. (65537/11 = 5957)
+			//	3.Allocate weight to each by multiple it’s original weight with number from step 2 (1*5957 for real one. 10 * 5957 for real 2)
```
Thanks for the extra detail. Is this comment meant to stay in the source or was this just to inform reviewers? It seems useful enough to keep, but we can strip out phrases like "So in your example on of em," and "Here at (fb)".

lrouter replied:
Thanks. You are right. I would keep this comment and strip out referenced phrases.

christarazi reviewed (changes requested):
Thanks for the PR. A few items to address below.

christarazi commented on pkg/maglev/maglev.go:
```diff
@@ -74,33 +75,60 @@ func getPermutation(backends []string, m uint64) []uint64 {
-	if len(backends) == 0 {
-		return nil
+// maglevBackendIDsBuffer. The slice contains backend IDs.
+func GetLookupTable(backends map[string]*BackendPoint, m uint64, maglevBackendIDsBuffer []uint16) {
```
Didn't catch this the first time around, but I noticed that we no longer return from this function anymore. Could you update the godoc accordingly?

lrouter replied:
OK. I have updated the descriptions.

Commit:
api: add maglev hash weight for backend point

Add maglev hash weight for backend in spec.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
service: add maglev weight field for service backend

Add maglev weight field for service backend. In order to test maglev weight, also support api to configure service backend weight.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
service: fix AddBackend definition error in LBMap interface

"LBMap" interface defines a function named "AddBackend". We have changed the implementation in lbmap file, but do not change the definition here. Now we fix it.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
bpf: add weight field in lb backend structure

We have added weight field in backend structure in user space golang code. Now we have add it in ebpf backend structures.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
backend_weight: update copyright date

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
lbmap: add UpsertMaglevLookupTable method

Add "UpsertMaglevLookupTable" method to take care of magelv algorithm calculation.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
service: update weight field of service

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>

Commit:
service: add backend weight in CiliumServiceAdd

Support backend weight for kubelet cli api.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
test: add maglev weight testcase

Add a maglev weight testcase with zero weight for a backend.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
test: fix service host ip

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
service_test: fix missing backend weight

Fix missing backend weight when create a new backend of lb package.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
service_update: panic when backend weights are not supplied

User should give backend weights when updating a service through cilium cli. Otherwise cli will panic.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
service: don't set backend weight default to 1

When backend weight is 0, do not set it default to 1.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

Commit:
service: set backend weight to 125 for echo service

The echo service has 2 backends, and maglev table size is 251. Choose 125 for backend weight to represent 0.5.

Signed-off-by: 范开喜 <fankaixi.li@bytedance.com>

oblazek commented on cilium/cmd/service_update.go:
```diff
@@ -148,15 +150,19 @@ func updateService(cmd *cobra.Command, args []string) {
 		}
 	}
 
+	if len(backendWeights) == 0 {
```
I suggest to also check whether user specified all backend weights, otherwise we might run into index out of range below
```suggestion
	if len(backendWeights) == 0 || len(backendWeights) != len(backends) {
```
